tools/engine_tool/lib/src/gn_utils.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:process_runner/process_runner.dart';
+
+import 'environment.dart';
+import 'proc_utils.dart';
+
+/// Information about a test build target.
+final class TestTarget {
+  /// Construct a test target.
+  TestTarget(this.label, this.executable);
+
+  /// The build target label.
+  final String label;
+
+  /// The executable file produced after the build target is built.
+  final File executable;
+
+  @override
+  String toString() {
+    return 'target=$label executable=${executable.path}';
+  }
+}
+
+/// Returns test targets for a given build directory.
+Future<Map<String, TestTarget>> findTestTargets(
+    Environment environment, Directory buildDir) async {
+  final Map<String, TestTarget> r = <String, TestTarget>{};
+  final List<String> getLabelsCommandLine = <String>[
+    'gn',
+    'ls',
+    buildDir.path,
+    '--type=executable',
+    '--testonly=true',
+    '--as=label',
+  ];
+  final ProcessRunnerResult labelsResult = await environment.processRunner
+      .runProcess(getLabelsCommandLine,
+          workingDirectory: environment.engine.srcDir, failOk: true);
+  fatalIfFailed(environment, getLabelsCommandLine, labelsResult);
+  final String rawLabels = labelsResult.stdout;
+  final List<String> getOutputsCommandLine = <String>[
+    'gn',
+    'ls',
+    buildDir.path,
+    '--type=executable',
+    '--testonly=true',
+    '--as=output'
+  ];
+  final ProcessRunnerResult outputsResults = await environment.processRunner
+      .runProcess(getOutputsCommandLine,
+          workingDirectory: environment.engine.srcDir, failOk: true);
+  fatalIfFailed(environment, getOutputsCommandLine, outputsResults);
+  final String rawOutputs = outputsResults.stdout;
+  final List<String> labels = rawLabels.split('\n');
+  final List<String> outputs = rawOutputs.split('\n');
+  if (labels.length != outputs.length) {
+    environment.logger.fatal(
+        'gn ls output is inconsistent A and B should be the same length:\nA=$labels\nB=$outputs');
+  }
+  // Drop the empty line at the end of the output.
+  if (labels.isNotEmpty) {
+    assert(labels.last.isEmpty);
+    assert(outputs.last.isEmpty);
+    labels.removeLast();
+    outputs.removeLast();
+  }
+  for (int i = 0; i < labels.length; i++) {
+    final String label = labels[i];
+    final String output = outputs[i];
+    assert(label.isNotEmpty);
+    assert(output.isNotEmpty);
+    r[label] = TestTarget(label, File(p.join(buildDir.path, output)));
+  }
+  return r;
+}
+
+/// Process selectors and filter allTargets for matches.
+///
+/// We support:
+///   1) Exact label matches (the '//' prefix will be stripped off).
+///   2) '/...' suffix which selects all targets that match the prefix.
```

Comment:
Running Windows tests with this syntax is a bit verbose:

```
et test flutter/shell/platform/windows/...
```

Some potential solutions:

1. Selectors can also be executable names. Windows tests become `et test flutter_windows_unittests`.
2. We support glob patterns. Windows tests become `et test **/windows/*`
3. We flatten the engine repo. Windows tests becomes `et test flutter/windows/...`

Thoughts?

Reply:
What Bazel supports is being in a sub-directory, in other words:

```sh
cd shell/platform/windows
et test :flutter_windows_unittests
```

I'd be supportive of supporting that in `et` (assuming GN is OK with it or it's easy to implement). I don't think we should create any other cute/creative short-hands.

Reply:
I just press "up" (or ctrl+R) in my terminal when I'm developing so I don't totally get the point that it's a lot to type.

I agree we should support the sub-directory workflow that Matan mentioned.

Also, if we get into a situation where we have to specify a bunch of subdirectories (et build/test subA/... subB/...) the fix might be reorganize the directory tree so that related code is near each other.